### PR TITLE
etcdctl/ctlv2: error handling with JSON

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/5970.

### CODE REVIEW

File: etcdctl/ctlv2/command/rm_command.go [hunk 1]
```diff
@@ -55,6 +58,15 @@ func rmCommandFunc(c *cli.Context, ki client.KeysAPI) {
...
 	cancel()
 	if err != nil {
+		switch c.GlobalString("output") {
+		case "json":
+			var index uint64
+			if resp != nil {
+				index = resp.Index
+			}
+			b, _ := json.Marshal(etcderr.NewError(etcderr.EcodeKeyNotFound, err.Error(), index))
+			err = fmt.Errorf("%s", b)
```
Comment (xiang90) [inline etcdctl/ctlv2/command/rm_command.go]: default: handleError(ExitServerError, err)

**Comment by heyitsanthony** [comment]
I think the output should be ` {"errorCode":100,"message":"Key not found","cause":"100: Key not found (/foo) [3]","index":0}`. The `Error:` bit makes the output non-json.

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 Now error prints out

```
./bin/etcdctl --debug --output json rm foo

{"errorCode":4,"message":"","cause":"100: Key not found (/foo) [4]","index":0}
```

for other commands as well

**Comment by xiang90** [comment]
seems like we should not keep the index field if it is always 0... this is pretty ugly hack, but i can live with it...

**Comment by gyuho** [comment]
Yeah should we change the `error.Error` json tag to omit zero int value when json-encoded?

**Comment by jlamillan** [comment]
@gyuho - were you just abbreviating the output or would we be really loosing the _debug_ output in an error scenario (your example shows --debug was active)?

Other than that, glad to see --output json working on error. Looking forward to this change.

**Comment by gyuho** [comment]
@jlamillan I don't think this change would affect `debug` output? /cc @xiang90

File: etcdctl/ctlv2/command/error.go [hunk 2]
```diff
@@ -30,10 +32,15 @@ const (
 	ExitClusterNotHealthy
 )
 
-func handleError(code int, err error) {
-	fmt.Fprintln(os.Stderr, "Error: ", err)
-	if cerr, ok := err.(*client.ClusterError); ok {
-		fmt.Fprintln(os.Stderr, cerr.Detail())
+func handleError(code int, err error, isJSON bool) {
+	if isJSON {
+		b, _ := json.Marshal(etcderr.NewError(code, err.Error(), 0))
```
Comment (xiang90) [inline etcdctl/ctlv2/command/error.go]: this is incorrect. code should be etcd error code, not exit code.

Comment (xiang90) [inline etcdctl/ctlv2/command/error.go]: do a type assertion against this https://github.com/coreos/etcd/blob/master/client/keys.go#L57 and simply marshal it to json and print it out

File: etcdctl/ctlv2/command/error.go [hunk 3]
```diff
@@ -30,10 +32,15 @@ const (
 	ExitClusterNotHealthy
 )
 
-func handleError(code int, err error) {
-	fmt.Fprintln(os.Stderr, "Error: ", err)
-	if cerr, ok := err.(*client.ClusterError); ok {
-		fmt.Fprintln(os.Stderr, cerr.Detail())
+func handleError(code int, err error, isJSON bool) {
```
Comment (xiang90) [inline etcdctl/ctlv2/command/error.go]: change this to handleError(c c *cli.Context, exitCode int, err error)

check if we need json format inside handleError. current code is too messy.

File: etcdctl/ctlv2/command/error.go (hunk 2 above)
Comment (gyuho) [inline etcdctl/ctlv2/command/error.go]: @xiang90 What do we do if the type-assert on `client.Error` fails?

Comment (xiang90) [inline etcdctl/ctlv2/command/error.go]: print out non json error. it is not an error from etcd.

Comment (gyuho) [inline etcdctl/ctlv2/command/error.go]: Ok got it. Thanks!

**Comment by gyuho** [comment]
@xiang90 Fixed to print out json only when the error typs is `*client.Error`. PTAL.

Thanks.

File: etcdctl/ctlv2/command/error.go [hunk 4]
```diff
@@ -30,7 +32,16 @@ const (
 	ExitClusterNotHealthy
 )
 
-func handleError(code int, err error) {
+func handleError(c *cli.Context, code int, err error) {
+	if c.GlobalString("output") == "json" {
+		if err, ok := err.(*client.Error); ok {
+			err.Code = code
+			b, _ := json.Marshal(err)
```
Comment (xiang90) [inline etcdctl/ctlv2/command/error.go]: get the error panic if there is an error.

**Comment by xiang90** [comment]
LGTM

**Pull request merged**